### PR TITLE
Enable proximity interrupt support for vcnl4010

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: script/vcnl4010/vcnl4010.lua [hunk 1]
```diff
@@ -117,28 +142,69 @@ local function vcnl4010:read_ambient()
     end
 end
 
+function vcnl4010:set_proximity_sample_rate(rate)
+    print(rate)
+
+    local status, _, reason = i2c.txn(i2c.tx(self.address, self.PROXIMITY_RATE, rate))
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: You can do multiple tx's in the same transaction right?

File: script/vcnl4010/vcnl4010.lua [hunk 2]
```diff
@@ -117,28 +142,69 @@ local function vcnl4010:read_ambient()
...
+    local status, _, reason = i2c.txn(i2c.tx(self.address, self.COMMAND, 1))
+    local status, _, reason = i2c.txn(i2c.tx(self.address, self.COMMAND, 3))
+    return status, reason
+end
+
+function vcnl4010:configure_interrupt(sample_type, low, high, count)
+    if sample_type == "proximity" then
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: Make these constants in the main table

File: script/vcnl4010/vcnl4010.lua [hunk 3]
```diff
@@ -117,28 +142,69 @@ local function vcnl4010:read_ambient()
...
+        sample_type = 0
+    elseif sample_type == "ambient" then
+        sample_type = 1
+    else
+        return false, "invalid sample type; must be ambient or proximity"
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: This is not recoverable so should be an assert right?

File: script/vcnl4010/vcnl4010.lua [hunk 4]
```diff
@@ -117,28 +142,69 @@ local function vcnl4010:read_ambient()
...
+    if not status then
+        return false, "failed to set low threshold register"
+    end
+    
+    status, _, reason = i2c.txn(i2c.tx(self.address, self.HIGH_THRESHOLD, string.pack(">I2", high)))
+    if not status then
+        return false, "failed to set high threshold register"
+    end
+
+    print(count, sample_type)
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: Remove these prints?

File: script/vcnl4010/vcnl4010.lua [hunk 5]
```diff
@@ -117,28 +142,69 @@ local function vcnl4010:read_ambient()
...
-    now = he.wait{time=now + SAMPLE_INTERVAL}
+    now, new_events, events = he.wait{time=now + SAMPLE_INTERVAL}
+    if new_events then
+        --print("interrupt!")
+        status, buffer, reason = i2c.txn(i2c.tx(sensor.address, sensor.PROXIMITYDATA), i2c.rx(sensor.address, 2))
+        dist = string.unpack(">I2", buffer)
+        print("TOO CLOSE", dist)
+        he.send_str("proximity", he.now(), "TOO CLOSE")
+        now, new_events, events = he.wait{time=he.now() + 1000} -- chill for a second
+        i2c.txn(i2c.tx(sensor.address, sensor.INTERRUPT_STATUS, 0xff)) -- clear the interrupt register
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: Doesn't this need to go outside of the if statement?

**Review by madninja — CHANGES_REQUESTED** [review]

File: script/vcnl4010/vcnl4010.lua (hunk 5 above)
Comment (Vagabond) [reply script/vcnl4010/vcnl4010.lua]: Only if we want to work around the 'interrupt bug' in the firmware?

**Comment by madninja** [comment]
Why yes we do..

File: script/vcnl4010/vcnl4010.lua (hunk 1 above)
Comment (Vagabond) [reply script/vcnl4010/vcnl4010.lua]: I'm not sure? Do we have documentation on this?

File: script/vcnl4010/vcnl4010.lua [hunk 6]
```diff
@@ -90,120 +109,174 @@ function vcnl4010:_get(reg, pack_fmt, convert)
...
+        return false, "unable to set value"
+    end
+    return newvalue
+end
+
 function vcnl4010:set_led_current(current)
     -- set the LED current, table 4
     if current < 0 or current > 20 then
-        return false, "VCNL4010 LED current value should be between 0 and 20"
+        assert(false, "VCNL4010 LED current value should be between 0 and 20")
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: You can omit the whole if statement.. 

`assert(current < 0 or current > 20, "...")`

File: script/vcnl4010/vcnl4010.lua [hunk 7]
```diff
@@ -90,120 +109,174 @@ function vcnl4010:_get(reg, pack_fmt, convert)
     return string.unpack(pack_fmt, buffer)
 end
 
+function vcnl4010:_update(reg, pack_fmt, update)
+    -- read register, apply a function to the result and write it back
+    -- number of bytes to read based on the pack string
+    if not update then
+        assert(false, "you must supply an update function")
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: You can omit the if..

`assert(update, "...")`

File: script/vcnl4010/vcnl4010.lua [hunk 8]
```diff
@@ -90,120 +109,174 @@ function vcnl4010:_get(reg, pack_fmt, convert)
...
         he.wait{time=1 + he.now()}
     end
 end
 
-function vcnl4010:set_proximity_sample_rate(rate)
-    print(rate)
+function vcnl4010:set_ambient_sample_rate(rate)
+    local status = self:_update(self.AMBIENT_RATE, "B", function(r) return r | (rate << 4) end)
+    if not status then
+        return false, "unable to set value"
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: be clearer in the message here? "failed to set ambient sample rate"

File: script/vcnl4010/vcnl4010.lua [hunk 9]
```diff
@@ -90,120 +109,174 @@ function vcnl4010:_get(reg, pack_fmt, convert)
...
+
+function vcnl4010:configure_interrupt(enable, sample_type, low, high, count)
+    if sample_type ~= self.PROXIMITY_INTERRUPT and
+       sample_type ~= self.AMBIENT_INTERRUPT then
+        assert(false, "invalid interrupt type; must be AMBIENT_INTERRUPT or PROXIMITY_INTERRUPT")
     end
 
     if count < self.INTERRUPT_COUNT_1 or count > self.INTERRUPT_COUNT_128 then
-        return false, "invalid interrupt count"
+        assert(false, "invalid interrupt count")
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: Put condition in assert

File: script/vcnl4010/vcnl4010.lua [hunk 10]
```diff
@@ -90,120 +109,174 @@ function vcnl4010:_get(reg, pack_fmt, convert)
...
+      function(r) return {proximity_ready=(r & self.PROXIMITY_READY_INTERRUPT) > 0,
+                          ambient_ready=(r & self.AMBIENT_READY_INTERRUPT) > 0,
+                          low_threshold=(r & self.THRESHOLD_LOW_INTERRUPT) > 0,
+                          high_threshold=(r & self.THRESHOLD_HIGH_INTERRUPT) > 0} end)
+end
+
+function vcnl4010:configure_interrupt(enable, sample_type, low, high, count)
+    if sample_type ~= self.PROXIMITY_INTERRUPT and
+       sample_type ~= self.AMBIENT_INTERRUPT then
+        assert(false, "invalid interrupt type; must be AMBIENT_INTERRUPT or PROXIMITY_INTERRUPT")
```
Comment (madninja) [inline script/vcnl4010/vcnl4010.lua]: Put condition in assert

**Review by madninja — CHANGES_REQUESTED** [review]

**Pull request merged**